core/src/main/java/org/apache/iceberg/io/PositionDeltaWriter.java

```diff
@@ -41,6 +41,21 @@
    */
   void insert(T row, PartitionSpec spec, StructLike partition);
 
+  /**
+   * Inserts a new version of an existing row to the provided spec/partition.
+   * <p>
+   * This method allows writers to distinguish new and updated records. The caller must separately
+   * invoke {@link #delete(CharSequence, long, PartitionSpec, StructLike)} for the original
+   * row position that is being updated.
+   *
+   * @param row a new version of an existing row
+   * @param spec a new partition spec
+   * @param partition a new partition or null if the spec is unpartitioned
+   */
+  default void update(T row, PartitionSpec spec, StructLike partition) {
```

Comment:
This needs to be discussed and I am open to alternatives. With the current change, the API seems a bit confusing but maybe we can play with names to make the purpose clear. In short, this method is to distinguish inserts from updates. In case of MERGE, we order updates by _spec_id, _partition, _file, _pos and updated records can be out of order if we change the sort key or partition columns. For inserts, the metadata columns are null so we have no problems ordering inserts.

Reply:
Splitting an update into two records (separate delete and separate insert) on the engine side is an option to consider but will require quite some effort to implement. Ideally, I would avoid that.

Reply:
If I understand correctly, the problem that this fixes is that the inserts are expected to be clustered, but we don't have the same expectation for updates because it's hard to come up with a single sort order that works for all cases (inserts, updates, and deletes).

So we separate updates that are not expected to be clustered and use a fanout writer for them?

Reply:
What do you think to have update take arguments (CharSequence path, long pos, T row, PartitionSpec spec, StructLike partition) and internally invoke delete()?  Maybe more fitting of the name, than having the caller call delete() first and then this method.

Reply:
Actually, both insert and update writers appear to be clustered. Do they just have a different clustering?

Reply:
I see. The partitioned writer has a comment about this. Updates come in so that deletes are clustered and ordered correctly -- by the path and position that will be deleted. But updates may have caused the row to change partitions and so that clustering for deletes may not match clustering for updates. So we do need to use a separate fanout writer.

Reply:
@szehon-ho, I thought about it but the original record may be in one spec/partition and the new record may be written to another one. As a result, we would need to have path, pos, row, oldSpec, oldPartition, newSpec, newPartition, which makes it a rather long list of arguments. I am happy to reconsider if folks prefer that API instead of this one.

@rdblue, correct, your last comment is spot on.

Reply:
Yea I see, that is unfortunately a bit of arguments, though still less than max limit of checkstyle.  Yea in my opinion, one single method is much cleaner than having the client remembering to call this one after delete(), but it is definitely a style choice and also not sure if anyone else agrees (@rdblue , @RussellSpitzer @jackye1995 ). Else we might  rename the method to something a bit awkward like 'updateDeleted' to be more clear.

Reply:
Let me create an issue where we can discuss these two options to avoid blocking the subsequent MERGE PR.

Reply:
Created #4042.

Reply:
Sure no problem, don't let it block